pilot/pkg/model/sidecar.go

```diff
@@ -216,25 +216,55 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *Config, configNamespa
 
 	// Assign namespace dependencies
 	out.namespaceDependencies = make(map[string]struct{})
+
+	addService := func(s *Service) {
+		if s == nil {
+			return
+		}
+		if foundSvc, found := servicesAdded[string(s.Hostname)]; !found {
+			servicesAdded[string(s.Hostname)] = s
+			out.services = append(out.services, s)
+			out.namespaceDependencies[s.Attributes.Namespace] = struct{}{}
+		} else if foundSvc.Attributes.Namespace == s.Attributes.Namespace && s.Ports != nil && len(s.Ports) > 0 {
+			// merge the ports to service when each listener generates partial service
+			// we only merge if the found service is in the same namespace as the one we're trying to add
+			os := servicesAdded[string(s.Hostname)]
+			for _, p := range s.Ports {
+				found := false
+				for _, osp := range os.Ports {
+					if p.Port == osp.Port {
+						found = true
+						break
+					}
+				}
+				if !found {
+					os.Ports = append(os.Ports, p)
+				}
+			}
+		}
+	}
+
 	for _, listener := range out.EgressListeners {
+		// First add the explicitly requested services
 		for _, s := range listener.services {
-			if _, found := servicesAdded[string(s.Hostname)]; !found {
-				servicesAdded[string(s.Hostname)] = s
-				out.services = append(out.services, s)
-				out.namespaceDependencies[s.Attributes.Namespace] = struct{}{}
-			} else if s.Ports != nil && len(s.Ports) > 0 {
-				// merge the ports to service when each listener generates partial service
-				os := servicesAdded[string(s.Hostname)]
-				for _, p := range s.Ports {
-					found := false
-					for _, osp := range os.Ports {
-						if p.Port == osp.Port {
-							found = true
-							break
-						}
-					}
-					if !found {
-						os.Ports = append(os.Ports, p)
+			addService(s)
+		}
+
+		// Infer more possible destinations from virtual services
+		// Services chosen here will not override services explicitly requested in listener.services.
+		// That way, if there is ambiguity around what hostname to pick, a user can specify the one they
+		// want in the hosts field, and the potentially random choice below won't matter
+		for _, vs := range listener.virtualServices {
+			v := vs.Spec.(*networking.VirtualService)
+			for _, d := range virtualServiceDestinations(v) {
+				// Default to this hostname in our config namespace
+				if s, ok := ps.ServiceByHostnameAndNamespace[host.Name(d.Host)][configNamespace]; ok {
+					addService(s)
+				} else {
+					// Otherwise pick randomly
+					for _, v := range ps.ServiceByHostnameAndNamespace[host.Name(d.Host)] {
+						addService(v)
+						break
```

Comment:
Couple of things come to my mind now:
1. Non determinism. Yes we are randomly picking from one of the hosts, but we want that choice to be constant because this code is called for every push epoch. So, if the choice of the service is different in each epoch, the listener or the route will change causing drains in Envoy (or worse, thrashing). So, you might want to get the list of namespaces, sort in stable manner and pick the (same) first one.
1. Virtual Service destination has a port. In the current model, you are simply adding the entire service here (including all the ports), while the logic that @ramaraochavali added a while ago was to explicitly filter out ports. Imagine a scenario like this:

```yaml
Sidecar:
egress:
- port: 9080
   hosts: ns1/scooby.com
- // wildcard catch all egress
  hosts: ns2/scrappy.com (used in virtual service)
---
VirtualService
host: Scrappy.com
namespace: ns2
route:
- destination: scooby.com
   port: 7080
---
ServiceEntry
namespace: ns1
host: scooby.com
ports: 8080, http
endpoints: scooby.corp.com
---
ServiceEntry
namespace: ns2
host: scooby.com
ports: 7080, http
endpoints: scooby.dev.com
---
```

Several things to note here:
  a. do we allow this kind of nightmarish configuration? No.
  b. If this kind of configuration comes through, what do we do? We need a deterministic way of resolving this. I dont thikn we need to support this scenario (I really hope not). Today, even without the virtual service import, I think we probably add the 7080 port to the ns1/scooby.com service [this is incorrect] and end up routing traffic from 7080 to ns1/scooby.com even though such a port does not exist. If we are not doing this, then great! If we are, then we should fix to disallow this @ramaraochavali / @howardjohn 
 I think this is where the logic that @jackkleeman added about namespace matching (if foundSvc.namespace == s.namespace) makes most sense. So, we should clearly document why (may be even add the contrived example I added above to the comments so that we remember why we did this in future).

Reply:
can you point me to where we are actually filtering service ports? i can see the logic to merge  but `selectServices` adds the full service afaict

Reply:
would it be sensible to filter ports from the destination of the vs too? maybe can be punted as a later optimization

Reply:
@rshriram Thanks for the detailed notes. 
Regarding case 2, that  you brought up, yes with today's code it incorrectly merges both the ports in `s1/scooby.com` and the namespace check will definitely fix that issue.

The changes related to sorting also look good to me.

However, I think the best thing to do here is apart from the code comments, add additional unit tests to cover the following scenarios
1. Same host across multiple namespaces with different ports without vs import and validate that port merging does not happen.
2. Multiple calls to createSidecar Scope returns the same host
3. When host does not exist across all namespaces

I agree, it looks like we can intelligently pick the host based on port match if available but we can do it as follow-up.

Reply:
Thinking about the sorting to pick a random one, the namespace sorting might also break if a new namespace comes with the same hostname - potentially a very rare thing, so config change might be tolerable. 

I am still wondering whether we should ignore if not available in `confignamespace` then we can explain it much better in docs also. And users can (and should) always be explicit in config when they have such conflicts. What is the advantage of we picking some host? It might help for those who have only one host across all namespaces - right?

@jackkleeman Does ignoring address your use case?

Reply:
No, in my use case the services can be in any namespace, but they won't have ambigous hostnames at its in k8s

Reply:
I can add those unit tests, no problem. Its definitely important to me that its possible to select a host from a different namespace, but I don't mind too much what mechanism we use to select, as there will only be one option for me

Reply:
Id be okay with only picking one if its unambigous, ie theres only one option, but I wonder is this preferable to picking arbitrarily?

Reply:
Ok. Got it. The advantage I see with picking only if there is no ambiguity(i.e. bynamespace len is exactly one) is there is  behaviour is deterministic. Any thing arbitrary might lead to runtime failures when config/services change. 

@rshriram @howardjohn what do you think? I am OK to move forward with this if you think random is fine in case of ambiguity.